### PR TITLE
HDDS-8591. Create scheduler to check for new root ca certificates

### PR DESCRIPTION
**Pull request by Galsza**
## What changes were proposed in this pull request?

After the SCMs finished their part of root ca rotation the clients also need to get the new root CA certificate. This is done through a simple polling mechanism that asks the SCMs for the root CAs and once it observes a change it invokes a consumer from the clients to notify them. Integrating this change is going to be done in later items.

## What is the link to the Apache JIRA

[HDDS-8591](https://issues.apache.org/jira/browse/HDDS-8591)

## How was this patch tested?

Added unit tests. Functional change is not yet observed in this patch as the polling is not actually integrated into anything yet.

### CODE REVIEW

**Comment by Galsza** [comment]
@ChenSammi, @fapifta please take a look at this patch

File: hadoop-hdds/common/src/main/resources/ozone-default.xml
Comment (ashishkumar50) [inline hadoop-hdds/common/src/main/resources/ozone-default.xml:2258]: `ozone.scm.security.service.address` This is duplicate config, already there is one. Don't see used in the new code, may be by mistake added. Please remove it.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/RootCaRotationPoller.java
Comment (ashishkumar50) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/RootCaRotationPoller.java:102]: Please correct log wrt cert and not state machine.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/RootCaRotationPoller.java
Comment (ashishkumar50) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/RootCaRotationPoller.java:72]: Would be nice to have a log here, when there is a new cert?

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/RootCaRotationPoller.java
Comment (ashishkumar50) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/RootCaRotationPoller.java:69]: Objects(`X509Certificate`) inside `knownRootCerts` and `scmRootCaCerts` are same when both have same content? Can you verify if it is fine. If not it will always enter into this condition.

**Review by ashishkumar50 — COMMENTED** [review]
@Galsza, Thanks for working on this. Please find few comments.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/RootCaRotationPoller.java
Comment (Galsza) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/RootCaRotationPoller.java]: Two certs are the same when their encoded content is the same. I think this should be good enough. I'm going to discuss this with @fapifta just to be sure and make a local test for it.

**Comment by Galsza** [comment]
@ashishkumar50 Please take another look!

File: hadoop-hdds/framework/src/test/java/org/apache/hadoop/hdds/security/x509/certificate/utils/TestRootCaRotationPoller.java
Comment (ashishkumar50) [inline hadoop-hdds/framework/src/test/java/org/apache/hadoop/hdds/security/x509/certificate/utils/TestRootCaRotationPoller.java:97]: Can you have small test by just adding knowCert to ensure both cases are fine.

Comment (Galsza) [reply hadoop-hdds/framework/src/test/java/org/apache/hadoop/hdds/security/x509/certificate/utils/TestRootCaRotationPoller.java:102]: I'm not sure what you meant by this, but I have refined the first test case to use the knownCert instead of not using certs at all. I hope this fixes it.

Comment (ashishkumar50) [reply hadoop-hdds/framework/src/test/java/org/apache/hadoop/hdds/security/x509/certificate/utils/TestRootCaRotationPoller.java:102]: Yes I meant the same. Thanks for updating it.

**Review by fapifta — COMMENTED** [review]
Thank you @Galsza for the work on this one, as we discussed today, let's wait with merging this one until we can get #4943 in, and then adjust this and the subsequent ones.

+1 from my end.

**Comment by ashishkumar50** [comment]
Thanks @Galsza, LGTM +1.

**Comment by Galsza** [comment]
Moving to draft until Sammi's patch is merged: https://github.com/apache/ozone/pull/4943

**Comment by Galsza** [comment]
Rebased to Sammi's change, since https://github.com/apache/ozone/pull/4943 is merged now

**Comment by Galsza** [comment]
@fapifta @adoroszlai @ChenSammi this pr is ready now, please take another look!

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/RootCaRotationPoller.java
Comment (ChenSammi) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/RootCaRotationPoller.java:142]: We can use stream function to achieve the join, like this 
certs.stream().map(X509Certificate::getSerialNumber).map(BigInteger::toString).collect(Collectors.joining(", "));

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/RootCaRotationPoller.java
Comment (ChenSammi) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/RootCaRotationPoller.java:132]: WARN instead of ERROR is more appropriate here,  Line 129 too.  Here is an example,

  LOG.warn("{} couldn't be stopped gracefully", getClass().getSimpleName());

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/RootCaRotationPoller.java
Comment (ChenSammi) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/RootCaRotationPoller.java:133]: We can avoid call the executor.shutdownNow() here, for shutdown is already called, and without the explicitly check isInterrupted() or any wait() like function, shutdownNow will have the same effect as shutdown.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/RootCaRotationPoller.java
Comment (ChenSammi) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/RootCaRotationPoller.java:102]: Error while trying to rotate root ca certificate 
->
Error while trying to poll root ca certificates

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/RootCaRotationPoller.java
Comment (ChenSammi) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/RootCaRotationPoller.java:67]: Better use "this.getClass().getSimpleName()" to avoid hard code class name in case it's changed later.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/HddsConfigKeys.java
Comment (ChenSammi) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/HddsConfigKeys.java:226]: I would suggest change the name to "hdds.x509.rootca.certificate.polling.interval".

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/security/SecurityConfig.java
Comment (Galsza) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/security/SecurityConfig.java]: @ChenSammi thanks I've addressed it now

Comment (ChenSammi) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/security/SecurityConfig.java]: Okay, let's wait for the CI to pass.

**Review by ChenSammi — APPROVED** [review]
The last patch LGTM +1.

**Pull request merged**
